### ISSUE TITLE
Feature: Formatter

### ISSUE THREAD

**Issue by yujinakayama** [rubocop/rubocop#222]
I'm thinking about adding formatter feature, much like RSpec's one.
This enables custom formatted output and more efficient integration with external programs.
https://www.relishapp.com/rspec/rspec-core/v/2-6/docs/command-line/format-option
https://www.relishapp.com/rspec/rspec-core/v/2-6/docs/formatters/custom-formatters

Also this would help #186.
Actually I have implemented most of this, though it's still wip and the base `master` is old.
https://github.com/yujinakayama/rubocop/commits/formatter
### Plan of Implementation
- Create skeleton formatter class `BaseFormatter` which implements public APIs
- Rework current `Report` classes into formatters
- Add `-f`, `--format` option to specify formatter (can be specified multiple times to allow multiple formatters)
- Deprecate `-e`, `--emacs` option (Probably it should be kept until RuboCop 1.0, because some external programs depend on this option)
- Add `-o`, `--output` option to specify output file (applies to the previously specified `--format`)
- Add `-r`, `--require` option to inject third-party formatter class into RuboCop
### Plan of Formatter Public APIs

I was thinking about also adding `#offence_detected`, but probably it's hard if parallel rubocop is introduced.

``` ruby
module Rubocop
  module Formatter
    class BaseFormatter
      attr_reader :output

      # @param output [IO]
      #   stdout or file
      def initialize(output)
        @output = output
      end

      # This method is invoked once before any files are processed.
      #
      # @param all_files [Array(String)]
      #   all file paths to be processed
      def started(all_files)
      end

      # This method is invoked at the beginning of processing each files.
      #
      # @param file [String]
      #   file path
      #
      # @param options [Hash]
      #   file specific information
      #   e.g. config for the file, disabled lines for each cops
      #   (probably this will be empty hash with initial implementation)
      def file_started(file, options)
      end

      # This method is invoked at the end of processing each files.
      #
      # @param file [String]
      #   file path
      #
      # @param offences [Array(Rubocop::Cop::Offence)]
      #    all detected offences for the file
      def file_finished(file, offences)
      end

      # This method is invoked after all files are processed,
      # or interrupted by user.
      #
      # @param processed_files [Array(String)]
      #   processed file paths
      def finished(processed_files)
      end
    end
  end
end

```

Any comments would be appreciated.
